apis/core/v1alpha1/hostedclusterpackage_types.go

```diff
@@ -0,0 +1,121 @@
+)
+
+// HostedClusterPackage defines package to be rolled out on every HyperShift HostedCluster.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,shortName=hcpkg
+// +kubebuilder:printcolumn:name="Available",type=string,JSONPath=`.status.conditions[?(@.type=="Available")].status`
+// +kubebuilder:printcolumn:name="Progressing",type=string,JSONPath=`.status.conditions[?(@.type=="Progressing")].status`
+// +kubebuilder:printcolumn:name="Generation",type=string,JSONPath=`.status.observedGeneration`
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+type HostedClusterPackage struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   HostedClusterPackageSpec   `json:"spec,omitempty"`
+	Status HostedClusterPackageStatus `json:"status,omitempty"`
+}
+
+// HostedClusterPackageSpec is the description of a HostedClusterPackage.
+type HostedClusterPackageSpec struct {
+	Strategy HostedClusterPackageStrategy `json:"strategy,omitempty"`
+	// HostedClusterSelector is a label query matching HostedClusters that the Package should be rolled out to.
+	HostedClusterSelector metav1.LabelSelector `json:"hostedClusterSelector,omitempty"`
+	// PackageSpec describes the Package that should be created when new
+	// HostedClusters matching the hostedClusterSelector are detected.
+	PackageSpec PackageSpec `json:"spec"`
+}
+
+// HostedClusterPackageStrategy describes the rollout strategy for a HostedClusterPackage.
+type HostedClusterPackageStrategy struct {
+	// MaxUnavailable defines how many Packages may become unavailable during upgrade at the same time.
+	// Cannot be below 1, because we cannot surge to create more instances.
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+	MaxUnavailable int `json:"maxUnavailable,omitempty"`
+	// Partition HostedClusters by label value.
+	// All packages in the same partition will have to be upgraded
+	// before progressing to the next partition.
+	Partition *HostedClusterPackagePartitionSpec `json:"partition,omitempty"`
+}
+
+// HostedClusterPackagePartitionSpec describes settings to partition HostedClusters into groups for upgrades.
+// Partitions will be upgraded after each other:
+// Upgrades in the next partition will only start after the previous has finished.
+type HostedClusterPackagePartitionSpec struct {
+	// LabelKey defines a labelKey to group objects on.
+	// +kubebuilder:validation:MinLength=1
+	LabelKey string `json:"labelKey"`
+	// Controls how partitions are ordered.
+	// By default items will be sorted AlphaNumeric ascending.
+	Order *HostedClusterPackagePartitionOrderSpec `json:"order,omitempty"`
+}
+
+// HostedClusterPackagePartitionOrderSpec describes ordering for a partition.
+type HostedClusterPackagePartitionOrderSpec struct {
+	// Allows to define a static partition order.
+	// The special * key matches anything not explicitly part of the list.
```

Comment:
How does ordering work for `*` matched items? Should this be communicated to the user here?